src/Adapter/MSTest.CoreAdapter/Execution/TypeCache.cs

```diff
@@ -635,9 +636,47 @@ private TestMethodAttribute GetTestMethodAttribute(MethodInfo methodInfo, TestCl
         /// <returns> The <see cref="MethodInfo"/>. </returns>
         private MethodInfo GetMethodInfoForTestMethod(TestMethod testMethod, TestClassInfo testClassInfo)
         {
-            var methodsInClass = testClassInfo.ClassType.GetRuntimeMethods().ToArray();
             MethodInfo testMethodInfo;
 
+            if (testMethod.HasManagedMethodAndType)
+            {
+                var assembly = testClassInfo.ClassType.GetTypeInfo().Assembly;
+                var methodBase = ManagedNameHelper.GetMethod(assembly, testMethod.ManagedTypeName, testMethod.ManagedMethodName);
+
+                if (methodBase is MethodInfo mi)
+                {
+                    testMethodInfo = mi;
+                }
+                else
+                {
+                    var parameters = methodBase.GetParameters().Select(i => i.ParameterType).ToArray();
+                    testMethodInfo = methodBase.DeclaringType.GetRuntimeMethod(methodBase.Name, parameters);
+                }
+
+                testMethodInfo = testMethodInfo.HasCorrectTestMethodSignature(true) ? testMethodInfo : null;
+            }
+            else
+            {
+                testMethodInfo = this.GetMethodInfoForLegacyTestMethod(testMethod, testClassInfo);
```

Comment:
nit: perhaps call this `GetMethodInfoFromRuntimeMethods` so its clear what it does.